hw02-DIYList/src/main/java/ru/otus/hw02/DIYList.java

```diff
@@ -0,0 +1,216 @@
+package ru.otus.hw02;
+
+import java.util.*;
+
+public class DIYList<E> implements List<E> {
+
+    private Object[] elements = new Object[10];
+    private int size = 0;
+
+    private class ListItr implements ListIterator<E> {
+
+        private  int cursor = 0;
+        private int lastRet = -1;
+
+        public boolean hasPrevious() {
+            return cursor != 0;
+        }
+
+        public E next() {
+            if (cursor >= size)
+                throw new NoSuchElementException();
+            cursor++;
+            return (E) elements[lastRet = cursor - 1];
+        }
+
+        public void remove() {
+            if (lastRet < 0)
+                throw new IllegalStateException();
+
+            DIYList.this.remove(lastRet);
+            cursor = lastRet;
+            lastRet = -1;
+        }
+
+        @Override
+        public boolean hasNext() {
+            return cursor != size;
+        }
+
+        public int nextIndex() {
+            return cursor;
+        }
+
+        public int previousIndex() {
+            return cursor - 1;
+        }
+
+        public E previous() {
+            if ( cursor - 1 < 0 ) {
+                throw new NoSuchElementException();
+            }
+            return (E) elements[lastRet = cursor - 1];
+        }
+
+        public void set(E e) {
+                DIYList.this.set(cursor - 1, e);
+        }
+
+        public void add(E e) {
+            DIYList.this.add(cursor, e);
+            cursor++;
+            lastRet = -1;
+        }
+    }
+
+    private void indexCheck(int index) {
+        if (index >= size || index < 0) {
+            throw new IndexOutOfBoundsException();
```

Comment:
Здесь в исключение полезно еще написать, какой именно index передали, а какой вы ожидали.
```suggestion
            throw new IndexOutOfBoundsException(...);
```